File: Nov2020/28Nov/PairSums.cpp
/*
Problem Name: Pair Sums
Problem Link: https://www.hackerearth.com/practice/data-structures/hash-tables/basics-of-hash-tables/practice-problems/algorithm/pair-sums/
*/

#include<bits/stdc++.h>
using namespace std;
int main(){
    ios::sync_with_stdio(0);    
    cin.tie(0);             
    int n,k;
    cin>>n>>k;
    int A[n];
    for(int i=0;i<n;++i)
        cin>>A[i];

    sort(A,A+n);
    bool ans=false;
    int f=0,l=n-1;
    while(f < l){
        if(A[f] + A[l] < k)
            f++;
        else if(A[f] + A[l] > k)
            l--;
        else{
            ans=true;
            break;
        }
    }
    if(ans) cout<<"YES";
    else    cout<<"NO";
    return 0;
}

File: Nov2020/28Nov/PairSums.java
/*
Problem Name: Pair Sums
Problem Link: https://www.hackerearth.com/practice/data-structures/hash-tables/basics-of-hash-tables/practice-problems/algorithm/pair-sums/
*/
import java.util.*;
class PairSums {
    public static void main(String args[] ) throws Exception {
        Scanner in = new Scanner(System.in);
        int n = in.nextInt(), k = in.nextInt();
        int arr[] = new int[10000000];
        for(int i=0;i<n;i++)
            arr[in.nextInt()]++;
        boolean pairs = false;
        
        for(int i=0;i<=k;i++){
            if(arr[i] > 0){
                if((k - i == i)){
                    if(arr[i] >= 2)
                        pairs = true;
                }
                else if(arr[k-i] > 0)
                    pairs = true;
            }
            if(pairs)
                break;
        }
        if(pairs)
            System.out.println("YES");
        else
            System.out.println("NO");
    }
}
